### DISCUSSION TITLE
Moving back to C++ Exceptions

### DISCUSSION THREAD

**Discussion by stephenberry** [stephenberry/glaze#1388] [Ideas]
> [!Note]
> This is a long term discussion looking years into the future for Glaze. A change will not be made unless we can ensure determinism on the error path and achieve reasonable error handling performance with exceptions.

Glaze should perhaps move back to using C++ exceptions. After experimenting for the past couple years disabling exceptions in Glaze, it has become apparent that this was perhaps not a good choice, and results in code that is slower, more complex, and less safe.

This talk is extremely helpful for viewing C++ exceptions and how to use them on embedded systems: [C++ Exceptions for Smaller Firmware - Khalil Estell - CppCon 2024](https://www.youtube.com/watch?v=bY2FlayomlE)

## Exceptions are faster
When we originally moved way from using exceptions in Glaze we saw on average a 5% - 10% loss in performance. Exceptions allow a reduction in branching, less binary overhead for large projects, and code that is easier to inline without cost. Glaze has needed to implement more macros under the hood in order to maintain performance, which makes debugging more difficult.

Glaze already avoids `std::expected` in the internal logic for performance reasons and only uses it at the top level API.

## Debugging is easier with exceptions
IDEs can break upon an exception being thrown. However, error codes do not trigger IDE breaks at the error location, which makes debugging and finding errors more difficult.

## Exceptions produce less binary in many cases
See Khalil Estell's talk linked above

We have seen this when developing code for Glaze on compiler explorer, and noticed it from performance losses due to additional branching and error code checks.

## Exceptions often result in safer code
I have interacted with users of Glaze who use void casting `(void)write_json(...` or `std::ignore` to discard errors. In many cases (especially writing) errors are extremely rare and therefore this is fine 99% of the time. However, hiding the 1% case is extremely dangerous. Exceptions allow these corner case errors to be properly handled without having to think about them and without them complicating code that is written everywhere. Also, many libraries like the STL just terminate when an exception would be raised, which makes code much less safe.

I've experienced issues with not checking an expected return and just calling `.value()`. This turns into an exception when exceptions are enabled, but when they are disabled it becomes a termination. The need to always check if the expected state has a value adds another layer of potential errors and sometimes results in major issues that wouldn't have occurred if exceptions were used.

## Using exceptions works better with the STL
The Standard Template Library uses exceptions everywhere, and when `-fno-exceptions` is used exceptions are turned into terminations. The STL is extremely useful for developing cross platform, efficient code.

We have found that needing to require two separate approaches to error handling (exceptions and error codes) results in more complex code, more lines of code, and more binary that runs slower.

Right now there are a number of functions labeled `noexcept` that could call user defined types that throw exceptions. One could argue that this is just a labelling issue, but it is a side effect of trying to avoid exceptions.

## Using exceptions results in cleaner code
Using exceptions majorly reduces code bloat and the need for ugly macros.

`glz::prettify_json` is an example of where we just return a `std::string`, because if it is taking a Glaze generated buffer then it should never error, but it could error. However, this error isn't reported because it would be annoying for the API. Currently users can use the internal API, but this just goes to show the problem of not using exceptions.

Consider this large macro in Glaze for skipping whitespace. The only reason this can't be an inlined function is that we need to be able to return out of the function on error without creating an intermediate boolean that adds performance overhead and binary. We found that compilers like GCC can sometimes remove nested returns, but in practice there is a real performance overhead and additional code complexity that can be avoided with the macro. Exceptions would allow this to be a normal C++ function that can be debugged.
```c++
#define GLZ_SKIP_WS(RETURN)                                              \
   if constexpr (!Opts.minified) {                                       \
      if constexpr (Opts.null_terminated) {                              \
         if constexpr (Opts.comments) {                                  \
            while (whitespace_comment_table[uint8_t(*it)]) {             \
               if (*it == '/') [[unlikely]] {                            \
                  skip_comment(ctx, it, end);                            \
                  if (bool(ctx.error)) [[unlikely]] {                    \
                     return RETURN;                                      \
                  }                                                      \
               }                                                         \
               else [[likely]] {                                         \
                  ++it;                                                  \
               }                                                         \
            }                                                            \
         }                                                               \
         else {                                                          \
            while (whitespace_table[uint8_t(*it)]) {                     \
               ++it;                                                     \
            }                                                            \
         }                                                               \
      }                                                                  \
      else {                                                             \
         if constexpr (Opts.comments) {                                  \
            while (it < end && whitespace_comment_table[uint8_t(*it)]) { \
               if (*it == '/') [[unlikely]] {                            \
                  skip_comment(ctx, it, end);                            \
                  if (bool(ctx.error)) [[unlikely]] {                    \
                     return RETURN;                                      \
                  }                                                      \
               }                                                         \
               else [[likely]] {                                         \
                  ++it;                                                  \
               }                                                         \
            }                                                            \
            if (it == end) [[unlikely]] {                                \
               ctx.error = error_code::end_reached;                      \
               return RETURN;                                            \
            }                                                            \
         }                                                               \
         else {                                                          \
            while (it < end && whitespace_table[uint8_t(*it)]) {         \
               ++it;                                                     \
            }                                                            \
            if (it == end) [[unlikely]] {                                \
               ctx.error = error_code::end_reached;                      \
               return RETURN;                                            \
            }                                                            \
         }                                                               \
      }                                                                  \
   }
```

*This is a working document to track my thoughts on exceptions and will continue to be updated.*

**Comment by 20162026** [comment]
I'm an embedded engineer who recently stumbled upon Glaze and was about to use it in a project, but mandatory exceptions are kind of a red flag.
I do understand that using standard containers without exceptions is extremely difficult, but I would really like to keep some kind of noexcept way to use Glaze, even if it's slower e.g. using a fixed buffer with length checks instead of strings.

For me, the main drawback of exceptions is the time it takes to unwind them in case of a sad path. It is difficult (if even possible) to calculate the exact time it would take to process an exception, which makes it unfeasible in real-time code, where you prioritize deterministic behaviour above everything else.
On the other hand, I don't know how often someone would really need to parse JSON in time-critical code paths or how bad it really is to use a simple exception that could be replaced with jump by compiler.

It might be fine to use exceptions for extremely rare and unexpected cases, e.g., when memory allocation fails, but using them for control flow in JSON parsing/validation is not ideal

**Reply by stephenberry** [reply]
I agree with you: it is important to have the error case for a parser performant and deterministic.

If you watch Khalil's talk, he's working on a tool to solve the deterministic problem, and will hopefully get his library into GCC and Clang.

This will take time and I want to see if it is achievable before changing Glaze back to exceptions. I agree that determinism in knowing the duration of the exception calls is critical.

I don't want to make this change lightly and so this is very much long term planning and discussion. It would probably be two or more years until this change were to be released, and it would be done only when it is shown to be robustly supported on at least GCC and Clang on Arm and x86.

**Comment by arturbac** [comment]
I am working on few large projects, old ones use exceptions, new ones use std::expected/unexpected with [simple_enum::generic_error_category](https://arturbac.github.io/simple_enum/generic_error_category.html) with low level try catch on anything that code uses externally that may throw like some stl containers.
I must say after ~1 year of experience of expected/unexpected, that the new projects are easier to understand when there could be an error, easier to solve error cases and easier to handle errors.
exceptions in large project very soon become a throwing black box with difficult to dertermine when and what may throw.

And there is a catch with exceptions, on some level error is not necessarily an error on other level. So exceptions make such error very inefficient to handle as mentioned before, you need to catch explicit set of exceptions instead of just handing unexpected.
With expected/unexpected such error handling is easy, fast and smooth.
You always can add to public functions throwing ones with different names or in some other namespace along with noexcept ones returning expected.
And every one will be happy.
From my perspective of use I don't want exceptions thrown from glaze as they are not critical errors usually so I really don't want try catch around any glaze use ... Much better is std::error_code or expected.

**Comment by cout** [comment]
Has there been any development on this in the last 2 years?

I come from a world where we assume exceptions can be thrown (so we make a lot of use of RAII), but we try to avoid throwing exceptions at all cost in our fast path.  As @20162026 pointed out, unwinding the stack for an exception can be expensive.  Our code is not hard realtime, but our performance requirements that are similar to realtime (the difference being that we have some tolerance for long tails).

As a user, I like the way glaze is -- I can use `glz::ex` and get exceptions, while users who avoid exceptions can use the functions that return an error code.

On the other hand, json serialization has recently become a performance bottleneck for us.  If glaze can be made even faster by switching to exceptions instead of error codes, that is something very seriously worth considering.  But if most of the performance gain is in parsing, that is less beneficial to us.  For parsing, it's also more useful to parse the whole stream even with errors, rather than stopping at the first error (which is what throwing an exception would do).

**Reply by stephenberry** [reply]
@cout, Thanks for asking. My opinion has shifted some in the last 2 years. While exceptions often lead to faster non-error paths, the performance on erroring paths for exceptions is slower than Glaze's short circuiting approach with error codes. Because errors are expected in JSON (since they often come from untrusted sources or can be human written), we need the error path to be extremely fast. Additionally, using error codes allows us to use the exact same code for verification as we would use for parsing, this saves huge amounts of code duplication. Whereas, if we used exceptions we would be making verification significantly slower or need to have a duplicated codebase for this purpose.

> json serialization has recently become a performance bottleneck for us

If you could describe the type of data you are serializing I might be able to give some tips on how to use Glaze more optimally. The write path has very minimal overhead for error handling, so changing to exceptions wouldn't benefit write speed.

> For parsing, it's also more useful to parse the whole stream even with errors, rather than stopping at the first error

Glaze's erroring system short circuits, which means it returns at first error. This offers massive performance improvements if you want to discard bad inputs because you don't have to parse the rest of the data. But, it actually also improves the parsing performance in general. Short circuiting allows us to write a parsing path that only supports valid data, if the data were invalid it would require branches that handle invalid data, which would increase the binary, branch misses, and slow down parsing.

To make parsing faster consider using `std::string_view`, or turn on the `minified` option if you know your data is always minified. Or use `glz::lazy_json` if you only care about some data and don't care if there are errors elsewhere.

Feel free to share more of your bottlenecks if you'd like some feedback on optimization.

**Reply by kammce** [reply]
I'm going to +1 @stephenberry . As the person mentioned in the original post, even with the improvements I've planned to make with exceptions regarding their performance and code size, I don't think exceptions aren't the right tool for situations where errors are extremely common. The point brought up about the write path having very little overhead for error handling makes sense, because any sort of error that could occur would be a bug in the text renderer and probably not the data itself. And I could see it being the opposite on the read side.

**Comment by cout** [comment]
@stephenberry Can you elucidate when as a user I should expect a function to return `error_ctx` vs `expected`?  Also, what should be the idiomatic pattern for converting them to exceptions and get the return value otherwise?  I see that `glz::format_error` works with either one, which is convenient, but some guidelines for how to use it would be helpful.

A helper function to throw a correctly formatted exception on error (preferably a type that inherits from `std::runtime_error` and retains the original `error_ctx`) and return the expected value on success (or void for an `error_ctx`) would be convenient and would tidy up some of the exception-throwing code in `json_exceptions.hpp` and elsewhere.
